File: Data Structures & Algorithms/house-robber/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
        // let n = nums.length;
        // if(n===1) return nums[0];
        // for(let i = 2;i<nums.length;i++){
        //     nums[i] = nums[i] + nums[i-2];
        // }
        // return Math.max(nums[nums.length-1],nums[nums.length-2]);
        let rob1 = 0;
            let rob2 = 0;

        for (const n of nums) {
            const temp = Math.max(n + rob1, rob2);
            rob1 = rob2;
            rob2 = temp;
        }
        return rob2;
    }
    
}

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        length = len(nums)

        if length == 1:
            return nums[0]
        
        if length == 2:
            return max(nums[0],nums[1])

        dp_FH = [0] * (length-1)
        dp_FH[0] = nums[0]
        if length > 1: dp_FH[1] = max(nums[0],nums[1])
        for i in range(2,length-1):
            dp_FH[i] = max(nums[i]+dp_FH[i-2],dp_FH[i-1])
        
        dp_LH = [0] * length
        dp_LH[0] = 0
        if length > 1: dp_LH[1] = nums[1]
        for i in range(2,length):
            dp_LH[i] = max(nums[i]+dp_LH[i-2],dp_LH[i-1])

        return max(dp_FH[length-2] , dp_LH[length-1])

        